blog: building slack apps with wing (#919)
winglang/docsite: blog/2024-05-25-bring-slack.md

    ctx.channel.postMessage(message);
  }
});
```

> Note: Be sure to replace the `INBOX_PROCESSING_CHANNEL` with the name of the slack channel you want to post to.

We can see above the very first resource defined is a `cloud.Secret` which is used to store the Slack bot token. This is a secure way to store sensitive information in Wing. So before we can get started we need to create a Slack app and get the bot token to use in our Wing app.

## Creating a Slack App

1. Head over to the [Slack API Dashboard](https://api.slack.com/apps) and create a new app.
2. Select `Create from Scratch` and give your app a name and select the workspace you want to deploy it to.
![Creating from scratch](assets/2024-05-25-bring-slack/from-scratch.png)
3. For the sake of our tutorial, we will be building a bot that requires the following permissions:
  - `app_mentions:read`
  - `chat:write`
  - `chat:write.public`
to add these head over to the `OAuth & Permissions` section and add those permissions.
![Adding permissions](assets/2024-05-25-bring-slack/permissions.png)
4. Install the app to your workspace and copy the bot token.
![Install app](assets/2024-05-25-bring-slack/install-to-workspace.png)

> Once its installed you can copy the bot token and lets head back over to our Wing app.

Reviewer: ```suggestion
Once its installed you can copy the bot token and let's head back over to our Wing app.
```

## Configuring Slack Bot Token Secret

> Now that we have our bot token, we can add it to our application simply by running the `wing secrets` command and pasting the token when prompted:

Reviewer: ```suggestion
Now that we have our bot token, we can add it to our application by running the `wing secrets` command and pasting the token when prompted:
```

```bash
❯ wing secrets
1 secret(s) found

? Enter the secret value for slack-bot-token: [hidden]
```

Now that our bot token is stored our application is ready to run!

## Running the App Locally

To run the app locally we can use the Wing Console, which simulates the cloud environment on your local machine. To start the console run:

```bash
wing it
```
This will open a browser window showing the Wing Console. You should see something similar to this:

![Wing Console](assets/2024-05-25-bring-slack/wing-console.png)

> Now in order to see the app in action lets add some more code to our `main.w` file. We will just add a simple function that will make a new file each time the function is called.

Reviewer: ```suggestion
Now in order to see the Slack bot in action, let's add some more code to our `main.w` file. We will add a function that will make a new file each time it is called.
```

The following code can be appended to the `main.w` file:

```js
let counter = new cloud.Counter();

new cloud.Function(inflight () => {
  let i = counter.inc();
  inbox.put("file-{i}.txt", "Hello, Slack!");
}) as "Add File";
```

Once you save the file, the Wing Console will hot reload and you should now see a function resource we can play with that looks like this:

![Add File Function](assets/2024-05-25-bring-slack/add-file-function.png)

So now we can click on the `Add File` function and interact with it in the right side panel. Go ahead and invoke the function a few times.

![Invoke Function](assets/2024-05-25-bring-slack/invoke-function.png)

> And BAM!! You should now be seeing messages in your slack channel every time you invoke the function!

Reviewer 2: Slack

![Slack Messages](assets/2024-05-25-bring-slack/slack-messages.png)

## Enabling Events

> One thing we will notice is the slack application is supposed to support the ability to list the files in the inbox. This is done by mentioning the slack app and saying `list inbox`. 

Reviewer: ```suggestion
One thing we will notice is the Slack application is supposed to support the ability to list the files in the inbox. This is done by mentioning the Slack app and saying `list inbox`. 
```

> If you try this now you will see absolutely nothing happens :) this is because we need to enable events in our slack app.

Reviewer: ```suggestion
If you try this now you will see absolutely nothing happens :) —— this is because we need to enable events in our slack app.
```

> To do this head over to the `Event Subscriptions` section in the slack API dashboard and enable events. You will need to provide a URL for the slack API to send events to. Luckily Wing makes providing this URL easy with builtin support for tunneling.

Reviewer 2: Slack

> To do this head over to the `Event Subscriptions` section in the slack API dashboard and enable events. You will need to provide a URL for the slack API to send events to. Luckily Wing makes providing this URL easy with builtin support for tunneling.

Reviewer 2: I'm just wondering if it will be confusing to the reader when you say to head over to the `Event Subscriptions` section, but they can't because they need a URL first. Maybe something like "Before we head to the ..."

Because you then direct them to head there a couple of paragraphs later.

To get the tunnel URL go back to the Wing Console and `Open a tunnel for this endpoint`

![Open Tunnel](assets/2024-05-25-bring-slack/open-tunnel.png)

> After a moment the icon will change to a eye with a slash through it, now we can copy the url by right clicking the endpoint and selecting `Copy URL`

Reviewer: ```suggestion
After a moment the icon will change to a eye with a slash through it, now we can copy the url by right clicking the endpoint and selecting `Copy URL`.
```

> Next let's head over to the `Event Subscriptions` section in the slack API dashboard and paste the URL in the `Request URL` field. However, we need to append `slack/events` to the end of the URL. So it should look something like this:

Reviewer 2: Slack

> Next lets head over to the `Event Subscriptions` section in the slack API dashboard and paste the URL in the `Request URL` field. However we need to append `slack/events` to the end of the URL. So it should look something like this:

Reviewer: ```suggestion
Next let's head over to the `Event Subscriptions` section in the slack API dashboard and paste the URL in the `Request URL` field. However, we need to append `slack/events` to the end of the URL. So it should look something like this:
```

![Request URL](assets/2024-05-25-bring-slack/events-sub-url.png)

This should only take a few seconds to verify, and once its verified you can scroll down to the `Subscribe to Bot Events` section and add the `app_mention` event like so:

![Subscribe to Bot Events](assets/2024-05-25-bring-slack/event-subscription.png)

Lastly, don't forget to save your changes!

### Testing the event subscription

Now head back to your Slack channel where you received the messages earlier and mention your app and say `list inbox`. Our Slack apps may have different names so it wont be exactly the same as the example below:

![List Inbox](assets/2024-05-25-bring-slack/list-inbox.png)

You should now see a message in the channel with the files in the inbox!

> Awesome! You have now built a Slack app using Wing and tested it locally. Now lets deploy it to AWS!

Reviewer: ```suggestion
Awesome! You have now built a Slack app using Wing and tested it locally. Now let's deploy it to AWS!
```

## Deploying to AWS
Before we start you will need the following to follow along:
1. An AWS Account
2. AWS CLI installed and configured with the necessary permissions.

> 3. Terraform installed

Reviewer: Maybe we can see if the example works with OpenTofu

```suggestion
3. Terraform or OpenTofu installed
```

Author: Good idea, Ill try it out

Getting our code ready for AWS is as simple as running 2 commands. The first thing we need to do is prepare the `cloud.Secret` for the `tf-aws` platform. This is done by running the `wing secrets` command with the `--platform` flag:

```bash
❯ wing secrets --platform tf-aws             
1 secret(s) found

? Enter the secret value for slack-bot-token: [hidden]
Storing secrets in AWS Secrets Manager
Secret slack-bot-token does not exist, creating it.
1 secret(s) stored AWS Secrets Manager
```

> This will result in the same prompt as before, but this time the secret will be stored in AWS secrets manager.

Reviewer: ```suggestion
This will result in the same prompt as before, but this time the secret will be stored in AWS Secrets Manager.
```

> Next lets compile the application for `tf-aws`:

Reviewer: ```suggestion
Next let's compile the application for `tf-aws`:
```

To deploy the code run the following commands:

```bash
terraform -chdir=./target/main.tfaws init
terraform -chdir=./target/main.tfaws apply -auto-approve
```

This will begin the deployment process and should only take about a minute to complete (barring internet connection issues). The result will show an output that contains a URL for the API Gateway endpoint and look something like this:

```bash
Apply complete! Resources: 30 added, 0 changed, 0 destroyed.

Outputs:

App_Api_Endpoint_Url_E233F0E8 = "https://p9y42fs0gg.execute-api.us-east-1.amazonaws.com/prod"
App_Slack_Request_Url_FF26641D = "https://p9y42fs0gg.execute-api.us-east-1.amazonaws.com/prod/slack/events"
```

The last step we need to do is to copy that `App_Slack_Request_Url` and paste it into the `Request URL` field in the `Event Subscriptions` section in the Slack API dashboard. This will tell Slack to now send events to our deployed applications API Gateway endpoint.

You should see the URL verified in a few seconds.

> DONT FORGET TO SAVE YOUR CHANGES!

![Request URL AWS](assets/2024-05-25-bring-slack/change-events-to-AWS.png)

### Playing with the deployed app

> Lets first test adding a file to the inbox, which in AWS is an s3 bucket. Navigate to the S3 console in AWS and find the bucket with the name that contains `file-process-inbox` there will be some unique hashing to the end of it, for example my bucket was named: `file-process-inbox-c8419ccc-20240530151737187700000004`

Reviewer: ```suggestion
Let's first test adding a file to the inbox, which in AWS is an s3 bucket. Navigate to the S3 console in AWS and find the bucket with the name that contains `file-process-inbox` there will be some unique hashing to the end of it. For example, my bucket was named: `file-process-inbox-c8419ccc-20240530151737187700000004`
```

> Upload any file on your machine to this bucket, and you should see a message in your slack channel!

Reviewer 2: Slack

![AWS Upload Slack Messages](assets/2024-05-25-bring-slack/aws-upload.png)

> Lastly, lets test the `list inbox` command.

Reviewer: ```suggestion
Lastly, let's test the `list inbox` command.
```
